### ISSUE TITLE
BUG: wait_for never match the condition

### ISSUE THREAD

**Issue by Dudesons** [alekc/terraform-provider-kubectl#178]
Hi,

I had a code using the previous version of the provider:
```hcl
resource "kubectl_manifest" "certificate" {
  depends_on = [
    kubectl_manifest.issuer,
  ]

  yaml_body = templatefile(
    "${path.module}/files/certificate.tmpl.yaml",
    {
      name         = local.certificate_name
      service_name = var.service_name
      namespace    = var.namespace
      secret_name  = local.secret_name
    },
  )

  wait_for {
    field {
      key   = "status.conditions.[0].type"
      value = "Ready"
    }

    field {
      key   = "status.conditions.[0].status"
      value = "True"
    }
  }
}
```

This one was working but since the last release, the `wait_for` is never happy with the condition I saw new improvements in the last release (`2.1.0`)

So I adapt my code to : 
```hcl
resource "kubectl_manifest" "certificate" {
  depends_on = [
    kubectl_manifest.issuer,
  ]

  yaml_body = templatefile(
    "${path.module}/files/certificate.tmpl.yaml",
    {
      name         = local.certificate_name
      service_name = var.service_name
      namespace    = var.namespace
      secret_name  = local.secret_name
    },
  )

  wait_for {
    condition {
      type   = "Ready"
      status = "True"
    }
  }
}
```

but with this update `wait_for` is never happy with the condition.
An example of what I have in kubernetes:
```yaml
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"cert-manager.io/v1","kind":"Certificate","metadata":{"annotations":{},"labels":{"app.kubernetes.io/name":"inference-operator"},"name":"inference-operator-self-signed","namespace":"inference-system"},"spec":{"dnsNames":["inference-operator.inference-system.svc","inference-operator.inference-system.svc.cluster.local"],"issuerRef":{"kind":"Issuer","name":"inference-operator-self-signed-issuer"},"secretName":"inference-operator-self-sign-cert","subject":{"organizationalUnits":["inference-operator"]}}}
  creationTimestamp: "2024-08-06T14:01:45Z"
  generation: 1
  labels:
    app.kubernetes.io/name: inference-operator
  name: inference-operator-self-signed
  namespace: inference-system
  resourceVersion: "142838381"
  uid: 9ef5e07d-e77c-4294-a1ee-f4daf158e9d0
spec:
  dnsNames:
  - inference-operator.inference-system.svc
  - inference-operator.inference-system.svc.cluster.local
  issuerRef:
    kind: Issuer
    name: inference-operator-self-signed-issuer
  secretName: inference-operator-self-sign-cert
  subject:
    organizationalUnits:
    - inference-operator
status:
  conditions:
  - lastTransitionTime: "2024-08-06T14:01:45Z"
    message: Certificate is up to date and has not expired
    observedGeneration: 1
    reason: Ready
    status: "True"
    type: Ready
  notAfter: "2025-01-03T14:01:45Z"
  notBefore: "2024-10-05T14:01:45Z"
  renewalTime: "2024-12-04T14:01:45Z"
  revision: 2
```

**Comment by 9numbernine9** [comment]
I'm noticing similar behaviour as well. The simplest test case I can come up with that seems to show the problem is this (silly) example:

```yaml
resource "kubectl_manifest" "test_wait_for" {
  yaml_body = <<EOF
apiVersion: v1
kind: Namespace
metadata:
  name: test-wait-for
EOF

  wait_for {
    field {
      key = "status.phase"
      value = "Active"
    }
  }
}
```

This seems to work fine with `2.0.4`, but will get stuck with `kubectl_manifest.test_wait_for: Still creating...` for seemingly forever when using `2.1.0` of the provider. Curiously, while the provider is waiting for the condition to be met, I can edit the YAML of the resource in-cluster and change that condition to something silly and it will then cause the provider to progress. In the example above, while the provider is waiting I can edit the YAML for that `Namespace` in-cluster and change `status.phase` to be `Activeish` (i.e. not `Active`) and _then_ the provider will respond with `Creation complete`.

**Comment by alekc** [comment]
Thanks for reproduction code, I will have a look

**Comment by pavel-spacil** [comment]
Hello, just tried a `condition` in a `wait_for` with the version `2.1.0` and `2.1.0-beta1` and the provider does not wait for any condition and immediately ends.

Tested with CRD:
```yaml
apiVersion: rabbitmq.com/v1beta1
kind: RabbitmqCluster
metadata:
  name: test
  namespace: rabbitmq
```

and

```terraform
  wait_for {
    condition {
      type   = "ClusterAvailable"
      status = "True"
    }
    condition {
      type   = "ReconcileSuccess"
      status = "True"
    }
  }
```

**Comment by alekc** [comment]
So, couple of issues was causing the behaviour at least for ns. 

1. if the desired change happened on the first observation, resourceversion was the same as in filter causing lack of event firing. 
2. even with resourceversion filter removed, the event.type was Added not modified, thus the change being skipped. 

I've removed the filtering on resource and added event.type added to the list of observations. Resourceversion might need to be put back with a skip clause (resourceVersion>event.resourceVersion => skip) but thats for later on. 

If tests passes I will release 2.1.1 to address this immediate issue.

**Comment by pavel-spacil** [comment]
I've just tried the new version and the `condition` check still does not work for me. 🙁
The CRD is the same like above, pasting relevant TF debug logs here:

```
2024-10-17T10:29:29.596+0200 [INFO]  provider.terraform-provider-kubectl_v2.1.1: 2024/10/17 10:29:29 [DEBUG] rabbitmq/test Unstructed YAML: map[apiVersion:rabbitmq.com/v1beta1 kind:RabbitmqCluster metadata:map[name:test namespace:rabbitmq] spec:map[persistence:map[storage:20Gi storageClassName:rabbitmq] replicas:1]]: timestamp="2024-10-17T10:29:29.596+0200"
2024-10-17T10:29:29.596+0200 [INFO]  provider.terraform-provider-kubectl_v2.1.1: 2024/10/17 10:29:29 [DEBUG] rabbitmq/test Unstructed YAML: map[apiVersion:rabbitmq.com/v1beta1 kind:RabbitmqCluster metadata:map[name:test namespace:rabbitmq] spec:map[persistence:map[storage:20Gi storageClassName:rabbitmq] replicas:1]]: timestamp="2024-10-17T10:29:29.596+0200"
2024-10-17T10:29:29.597+0200 [WARN]  Provider "registry.terraform.io/alekc/kubectl" produced an invalid plan for kubectl_manifest.test, but we are tolerating it because it is using the legacy plugin SDK.
    The following problems may be the cause of any confusing errors from downstream operations:
      - .wait_for_rollout: planned value cty.True for a non-computed attribute
      - .server_side_apply: planned value cty.False for a non-computed attribute
      - .apply_only: planned value cty.False for a non-computed attribute
      - .field_manager: planned value cty.StringVal("kubectl") for a non-computed attribute
      - .force_new: planned value cty.False for a non-computed attribute
      - .validate_schema: planned value cty.True for a non-computed attribute
      - .force_conflicts: planned value cty.False for a non-computed attribute
2024-10-17T10:29:29.597+0200 [INFO]  Starting apply for kubectl_manifest.test
2024-10-17T10:29:29.597+0200 [DEBUG] kubectl_manifest.test: applying the planned Create change
2024-10-17T10:29:29.598+0200 [INFO]  provider.terraform-provider-kubectl_v2.1.1: 2024/10/17 10:29:29 [DEBUG] rabbitmq/test Unstructed YAML: map[apiVersion:rabbitmq.com/v1beta1 kind:RabbitmqCluster metadata:map[name:test namespace:rabbitmq] spec:map[persistence:map[storage:20Gi storageClassName:rabbitmq] replicas:1]]: timestamp="2024-10-17T10:29:29.598+0200"
2024-10-17T10:29:29.598+0200 [INFO]  provider.terraform-provider-kubectl_v2.1.1: 2024/10/17 10:29:29 [DEBUG] rabbitmq/test apply kubernetes resource:
apiVersion: rabbitmq.com/v1beta1
kind: RabbitmqCluster
metadata:
  name: test
  namespace: rabbitmq
spec:
  replicas: 1
  persistence:
    storageClassName: rabbitmq
    storage: 20Gi: timestamp="2024-10-17T10:29:29.598+0200"
2024-10-17T10:29:29.601+0200 [INFO]  provider.terraform-provider-kubectl_v2.1.1: 2024/10/17 10:29:29 [INFO] rabbitmq/test perform apply of manifest: timestamp="2024-10-17T10:29:29.601+0200"
2024-10-17T10:29:32.287+0200 [INFO]  provider.terraform-provider-kubectl_v2.1.1: 2024/10/17 10:29:32 [INFO] rabbitmq/test manifest applied, fetch resource from kubernetes: timestamp="2024-10-17T10:29:32.286+0200"
2024-10-17T10:29:32.309+0200 [INFO]  provider.terraform-provider-kubectl_v2.1.1: 2024/10/17 10:29:32 [DEBUG] rabbitmq/test fetched successfully, set id to: /apis/rabbitmq.com/v1beta1/namespaces/rabbitmq/rabbitmqclusters/test: timestamp="2024-10-17T10:29:32.309+0200"
2024-10-17T10:29:32.334+0200 [INFO]  provider.terraform-provider-kubectl_v2.1.1: 2024/10/17 10:29:32 [INFO] rabbitmq/test waiting for wait conditions for 10min: timestamp="2024-10-17T10:29:32.334+0200"
2024-10-17T10:29:32.356+0200 [INFO]  provider.terraform-provider-kubectl_v2.1.1: 2024/10/17 10:29:32 [TRACE] Received event type ADDED for test: timestamp="2024-10-17T10:29:32.356+0200"
2024-10-17T10:29:32.357+0200 [INFO]  provider.terraform-provider-kubectl_v2.1.1: 2024/10/17 10:29:32 [DEBUG] rabbitmq/test fetch from kubernetes: timestamp="2024-10-17T10:29:32.357+0200"
```
The conditions don't have desired values just after the creation, it takes ~1 minute to have them...

Should be this issue reopened or I should open a new issue for this? 

Thank you!

**Comment by alekc** [comment]
Argh. I will reopen this one and have a further look.

**Comment by alekc** [comment]
Sigh, this is a serious blunder. Long story short, during previous refactor a logical issue has been introduced where the value of wait_for field/condition is not being properly evaluated. Basically 
```
resource "kubectl_manifest" "test_wait_for" {
  timeouts {
    create = "20s"
  }
  yaml_body = <<EOF
apiVersion: v1
kind: Namespace
metadata:
  name: test-wait-for
EOF

  wait_for {
    field {
      key = "status.phase"
      value = "Active"
    }
  }
}`
``` 
will be true. Following:
```
resource "kubectl_manifest" "test_wait_for" {
  timeouts {
    create = "20s"
  }
  yaml_body = <<EOF
apiVersion: v1
kind: Namespace
metadata:
  name: test-wait-for
EOF

  wait_for {
    field {
      key = "status.phase"
      value = "ZZZ"
    }
  }
}`
``` 
Will pass as well.  

Working on the fix.

**Comment by alekc** [comment]
2.1.2 is being deployed. Please check if that fixes your issue.

**Comment by alekc** [comment]
Reclosing this off. If anyone is still suffering from the issue please let me know.

**Comment by stevehipwell** [comment]
@alekc am I correct in thinking that #180 fixed the way I switched over to using a watch for `wait_for` and #182 fixed the refactored `wait_for` logic from #157?

I've opened #190 to cleanup after #180 and to optimise the delete watch pattern.

**Comment by pavel-spacil** [comment]
@alekc sorry for a late response but I've just tried the new version and it works for me now. Thanks for a quick fix! 👍
